pkg/kubelet/images/helpers.go

```diff
@@ -50,3 +51,55 @@ func (ts throttledImageService) PullImage(ctx context.Context, image kubecontain
 	}
 	return "", fmt.Errorf("pull QPS exceeded")
 }
+
+// ParallelBlocker for parallel puller control.
+type ParallelBlocker interface {
+	// Wait if it needs block pull request.
+	// Recover is used to unblock after process.
+	Wait(image kubecontainer.ImageSpec) Recover
+}
+
+// Recover function of release lock of ParallelBlocker from Wait() was called.
+type Recover func()
+
+// defaultParallelBlocker creates a ParallelBlocker with default implementation blockerWithMaxParallel.
+// Blocker will not limit concurrency  with maxParallelImagePulls nil or less than 1.
+func defaultParallelBlocker(maxParallelImagePulls *int32) ParallelBlocker {
+	if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
+		return &blockerWithMaxParallel{
+			imagePullingMap:          sync.Map{},
+			tokensForDistinctImageID: nil,
+		}
+	} else {
+		return &blockerWithMaxParallel{
+			imagePullingMap:          sync.Map{},
+			tokensForDistinctImageID: make(chan struct{}, *maxParallelImagePulls),
+		}
+	}
+}
+
+// blockerWithMaxParallel implements ParallelBlocker as default.
+// limits the number of concurrent calls.
+type blockerWithMaxParallel struct {
+	imagePullingMap          sync.Map
+	tokensForDistinctImageID chan struct{}
+}
+
+// Wait limits with tokensForDistinctImageID
+// Image pull requests with the same imageID will not be executed concurrently.
+// This ensures that if a Node serving numerous Pods with the same image `X`
+// and one Pod with the different image `Y`, The same image name will be pulled serially,
+// instead of `Y` waiting on `maxParallelImagePulls`'s pulls of `X`.
+func (b *blockerWithMaxParallel) Wait(image kubecontainer.ImageSpec) Recover {
+	if b.tokensForDistinctImageID != nil {
+		tokensForSameImageID, _ := b.imagePullingMap.LoadOrStore(image.Image, make(chan struct{}, 1))
+		tokensForSameImageID.(chan struct{}) <- struct{}{}
+		b.tokensForDistinctImageID <- struct{}{}
```

Comment:
>The last iteration is not accounting even for a handler.
I think there's a misunderstanding here. Different from the previous logic, all images will be subject to concurrency restrictions and there will be no unlimited situation.
The only difference is that images with same image ID are queued **one more time** and will not occupy the maximum number of concurrent users at the same time.
A more pessimistic limit than before that does not rely on more assumptions.
Just solve “`Y` waiting on `maxParallelImagePulls`'s pulls of `X`” that problem. ☺
@SergeyKanzhelev @HirazawaUi

Reply:
I understand what this PR did. My comment is about the situation when two images with the same name needs to be downloaded because of a runtime handler or because of annotations. We will slow down this case.

I understand that the case you are optimizing for is more common. But I am worried that by placing this logic we are limiting future scenarios. For example, passing a containerd snapshotter to use as an annotation. So two images are different by annotation and being downloaded differently.

I think some sort of filtering annotations to not pass them to containerd would be the best solution here.